### PR TITLE
Inference metrics improvements

### PR DESCRIPTION
**Pull request by rfejgin**
Added these two metrics to `infer_and_evaluate.py`:
* Add MOS estimation using UTMOSv2
* Add a metric to tracking total generated audio duration per dataset. We can use this as a lightweight indicator of speaking rate.
* Rename a command line option for consistency: `disable_fcd` --> `--no_fcd`

### CODE REVIEW

File: scripts/magpietts/infer_and_evaluate.py [hunk 1]
```diff
@@ -295,6 +295,7 @@ def run_inference(
     violin_plot_metrics=['cer', 'pred_context_ssim'],
```
Comment (rfejgin) [inline scripts/magpietts/infer_and_evaluate.py:295]: I did not add UTMOS to the set of violin plots, should we though?  Would be interested in comments.

Pro: this way it'll actually be looked at more frequently
Con: the combined plot would get crowded with 3 violin plots per dataset. (Or maybe UTMOS can be added only to the single-dataset plots?)

File: .github/workflows/cicd-main-speech.yml [hunk 2]
```diff
@@ -194,6 +194,9 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v4
+      # Temporarily install this manually until we add it to the docker image.
+      - name: Install UTMOSv2 # Needed by evaluate_generated_audio.py.
+        run: pip install git+https://github.com/sarulab-speech/UTMOSv2.git
```
Comment (shehzeen) [inline .github/workflows/cicd-main-speech.yml]: should we use a specific commit or are we comfortable with using their latest main branch?

File: scripts/magpietts/evaluate_generated_audio.py [hunk 3]
```diff
@@ -408,7 +431,8 @@ def evaluate(
         hypotheses=gt_audio_texts, references=gt_texts, use_cer=False
     )[0]
     avg_metrics["frechet_codec_distance"] = fcd
-
+    avg_metrics["utmosv2_avg"] = sum([m['utmosv2'] for m in filewise_metrics]) / len(filewise_metrics)
+    avg_metrics["total_gen_audio_seconds"] = total_generated_audio_seconds
```
Comment (shehzeen) [inline scripts/magpietts/evaluate_generated_audio.py:450]: How about instead of this, we use phonemes/second? If it's challenging to get number of phonemes, we can use characters or words (split by spaces or special chars) instead.

File: scripts/magpietts/infer_and_evaluate.py (hunk 1 above)
Comment (shehzeen) [reply scripts/magpietts/infer_and_evaluate.py:295]: I think we should add it to the violin plots! It's a valuable metric to look at when comparing models.

File: .github/workflows/cicd-main-speech.yml (hunk 2 above)
Comment (rfejgin) [reply .github/workflows/cicd-main-speech.yml]: Yeah, I was wondering the same. But I looked at their recent commits and it's basically infrequent maintenance updates (e.g. compatibility with new versions of torch). Probably less than 1 commit a month on average. So I tend to think we benefit more from using latest on main branch since it should have bugfixes etc. but minimal churn from updates.

File: scripts/magpietts/infer_and_evaluate.py (hunk 1 above)
Comment (rfejgin) [reply scripts/magpietts/infer_and_evaluate.py:295]: Okay good — added! `UTMOSv2` is now included in the violin plots by default. The user can still choose a different set a metrics if desired via the `--violin_plot_metrics` option.

File: scripts/magpietts/evaluate_generated_audio.py (hunk 3 above)
Comment (rfejgin) [reply scripts/magpietts/evaluate_generated_audio.py:450]: Yeah, we started discussing a bit in the meeting and Jason was of the opinion we should start simple, with just total duration. @blisc - any further thoughts?

File: .github/workflows/cicd-main-speech.yml (hunk 2 above)
Comment (rfejgin) [reply .github/workflows/cicd-main-speech.yml]: Anyways: thought about it some more and decided to update pip-requirements to pin UTMOSv2 to its latest release, `v1.2.1`.

File: scripts/magpietts/evaluate_generated_audio.py (hunk 3 above)
Comment (rfejgin) [reply scripts/magpietts/evaluate_generated_audio.py:450]: Discussed this a bit with Jason - speech rate would be useful but we'll leave it for a future PR. This PR was initially meant to focus on UTMOS and we don't want to complicate the PR too much.

File: scripts/magpietts/infer_and_evaluate.py [hunk 4]
```diff
@@ -682,17 +686,18 @@ def main():
...
     parser.add_argument('--cer_target', type=float, default=None)
     parser.add_argument('--ssim_target', type=float, default=None)
+    parser.add_argument('--no-utmosv2', action='store_true', help="Disable UTMOSv2 computation")
     parser.add_argument(
         '--log_exp_name',
         action='store_true',
         help="Include the experiment name (derived from the checkpoint path) in the output folder name.",
     )
-    parser.add_argument('--disable_fcd', action='store_true', help="Disable Frechet Codec Distance computation")
+    parser.add_argument('--no_fcd', action='store_true', help="Disable Frechet Codec Distance computation")
```
Comment (blisc) [inline scripts/magpietts/infer_and_evaluate.py]: Stylistically, let's keep the older format of `disable_{metric}`

File: scripts/magpietts/infer_and_evaluate.py [hunk 5]
```diff
@@ -295,6 +295,7 @@ def run_inference(
     violin_plot_metrics=['cer', 'pred_context_ssim'],
     eos_detection_method=None,
     ignore_finished_sentence_tracking=False,
+    with_utmosv2=True,
```
Comment (blisc) [inline scripts/magpietts/infer_and_evaluate.py:298]: Since we know that this is a slow inference call, can we default this to be disabled? And enable it only for some smaller eval datasets?

File: nemo/collections/tts/modules/utmosv2.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            # without actually speeding up prediction. Limit to 4 threads.
+            with threadpool_limits(limits=4):
+                mos_score = self.model.predict(input_path=file_path, num_repetitions=1, num_workers=0)
+        return mos_score
+
+
+def get_available_device():
+    """
+    Get the best available device (prefer GPU, fallback to CPU).
+    """
```
Comment (blisc) [inline nemo/collections/tts/modules/utmosv2.py:77]: Let's refactor this. Can you include a device in UTMOSv2Calculator's init that defaults to None? In our script, we should pass the a device that makes sense to our scripts. If it is None, we can then call this function, but we shouldn't rely on this to decide where to place the eval model.

File: scripts/magpietts/evaluate_generated_audio.py [hunk 7]
```diff
@@ -334,6 +351,11 @@ def evaluate(
                 gt_context_ssim_alternate = torch.nn.functional.cosine_similarity(
                     gt_speaker_embedding_alternate, context_speaker_embedding_alternate, dim=0
                 ).item()
+            if with_utmosv2:
+                utmosv2_score = utmosv2_model(pred_audio_filepath)
```
Comment (blisc) [inline scripts/magpietts/evaluate_generated_audio.py]: Is there a reason that we want to run this at batch 1? This might require a larger refactor of the code though. so if we cannot do it in this PR, can we skip this comment

**Review by blisc — CHANGES_REQUESTED** [review]

File: scripts/magpietts/infer_and_evaluate.py (hunk 4 above)
Comment (rfejgin) [reply scripts/magpietts/infer_and_evaluate.py]: Done

File: scripts/magpietts/infer_and_evaluate.py (hunk 5 above)
Comment (rfejgin) [reply scripts/magpietts/infer_and_evaluate.py:298]: I've changed it so that if the dataset has more than 200 entries we skip UTMOS calculation.

File: nemo/collections/tts/modules/utmosv2.py (hunk 6 above)
Comment (rfejgin) [reply nemo/collections/tts/modules/utmosv2.py:77]: Done

File: scripts/magpietts/evaluate_generated_audio.py (hunk 7 above)
Comment (rfejgin) [reply scripts/magpietts/evaluate_generated_audio.py]: I had played around with that earlier on and the issue was the speed was mostly bottlenecked CPU-side processing (feature extraction). And it was tricky to tune the number of workers and OpenMP threads to get an actual speedup from batching.

But I played around it a bit more now and I think I found a faster configuration (with batching and a particular thread configuration). Still need to integrate the directory-based processing with the main evaluation script.

File: scripts/magpietts/evaluate_generated_audio.py [hunk 8]
```diff
@@ -153,6 +168,18 @@ def extract_embedding(model, extractor, audio_path, device, sv_model_type):
     return embeddings.squeeze()
 
 
+def compute_utmosv2_scores(audio_dir):
+    print(f"\nComputing UTMOSv2 scores for files in {audio_dir}...")
+    start_time = time.time()
+    utmosv2_calculator = UTMOSv2Calculator()
```
Comment (blisc) [inline scripts/magpietts/evaluate_generated_audio.py]: Can we use the same "device" as on line 200: `device = "cuda"`

Comment (rfejgin) [reply scripts/magpietts/evaluate_generated_audio.py]: Good catch! Fixed.

**Comment by rfejgin** [comment]
I removed the restriction limiting UTMOSv2 to datasets with under 200 utterances. That's because it's a lot faster to run now that it is batched and with inference workers and threads tuned. For a 2000-utterance dataset, UTMOSv2 only increased test time from ~25 min to 28 min which seems worth it.

Will merge once CI passes (excluding unrelated known CI issue with coverage calculation).
